### PR TITLE
Py3support

### PR DESCRIPTION
**Pull request by AdityaSripal**
Added python3 compatibility and fixed rebasing issue.

### CODE REVIEW

File: pandevice/__init__.py [hunk 1]
```diff
@@ -136,24 +143,45 @@ def subrelease_num(self):
...
-                return -1
-            elif self.subrelease_type != 'c' and other.subrelease_type == 'c':
-                return 1
-            elif self.subrelease is None and other.subrelease_type == 'b':
-                return 1
-
-        return cmp(self.version, other.version)
-
-
+    def __lt__(self, other):
```
Comment (shinmog) [inline pandevice/__init__.py]: The logic is flawed in here somewhere.  This is evaluating to `True` and should be `False`:

`PanOSVersion('6.1.0') < PanOSVersion('6.1.0')`

I think it would be worthwhile to add test cases around this class' changes to make sure that these evaluations are working as expected.  It will probably be a lot of tests, but that's ok.  Since pytest is being used as the test runner now, go ahead and use a parameterized test to make your life easier.

Comment (AdityaSripal) [inline pandevice/__init__.py]: I believe I have fixed the issue. I added a more extensive test file to check and it seems to be working. Will add more tests tomorrow.

File: tests/test_PanOScomp.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import pytest
```
Comment (shinmog) [inline tests/test_PanOScomp.py]: You're close on this one, but it needs some redesign.  Assuming you've fixed the bug I noticed from before, these tests will not catch any future regressions introduced to the PanOSVersion comparison code.

So, there are six comparisons, and thus six comparison functions you've developed:
1) ==
2) !=
3) \>
4) \>=
5) \<
6) \<=

Thus you need six different test sets to verify everything, one `def` per comparison operation under test.  What I would like to see is that your six different test `def`s would take in three arguments:  two strings to turn into `PanOSVersion`s and whether the comparison under test should succeed or fail.

File: pandevice/__init__.py [hunk 3]
```diff
@@ -66,10 +66,17 @@ def getlogger(name=__name__):
...
-    reverse = dict((value, key) for key, value in enums.iteritems())
+    reverse = dict(((v, k) for (k, v) in enums.items()))
     enums['reverse_mapping'] = reverse
     return type('Enum', (), enums)
 
+def isstring(arg):
+    try:
+        return isinstance(arg, basestring)
+    except NameError:
+        return isinstance(arg, str) or isinstance(arg, bytes)
```
Comment (AdityaSripal) [inline pandevice/__init__.py]: There are a few reasons why I am hesitant to do this, even though the tests still pass. I think its imperative that isstring() returns the same value given the same argument regardless of whether it runs on Python2 or Python3.
To do that I could either make isstring check if arg is a Python2 basestring (current implementation) or a Python3 string. I couldn't find a good way to check if arg is a Python3 string in Python2 because Python2 makes no distinction between byte literals and string literals.
Making the function only return true for basestring and str will yield different return values between Python2 and Python3 when passed in a bytes literal.

File: tests/test_integration.py [hunk 4]
```diff
@@ -167,6 +171,7 @@ def test_element_str_from_highavailability_with_ha1_and_ha2_children(self):
     # 2) VirtualRouter with StaticRoute child
     def test_element_str_from_virtualrouter_with_sr_parent(self):
         '''StaticRoute > VirtualRouter'''
+        self.maxDiff = None
```
Comment (AdityaSripal) [inline tests/test_integration.py]: When I was debugging the issue with the strings splitting up, I needed to see the full expected and return value not the clipped versions. I can delete that line if needed.

File: examples/userid.py [hunk 5]
```diff
@@ -106,7 +106,7 @@ def main():
         logging.debug("Logout user %s at IP %s" % (args.user, args.ip))
         device.userid.logout(args.user, args.ip)
     else:
-        raise StandardError("Unknown action: %s.  Must be 'login' or 'logout'." % args.action)
+        raise Exception("Unknown action: %s.  Must be 'login' or 'logout'." % args.action) #Standard Error no longer supported
```
Comment (btorresgil) [inline examples/userid.py]: Good catch, though this probably should have been a ValueError from the beginning.

File: pandevice/base.py [hunk 6]
```diff
@@ -1873,7 +1876,10 @@ def parse_xml(self, xml):
 
     def element_str(self):
         """The XML form of this object (and its children) as a string."""
-        return ET.tostring(self.element())
+        val = ET.tostring(self.element())
+        if hasattr(val, "hex"):
```
Comment (btorresgil) [inline pandevice/base.py]: Possible other way to handle this instead of checking for "hex" attribute:

    try:
        # Python3
        return ET.tostring(self.element(), encoding='unicode')
    except LookupError:
        # Python 2
        return ET.tostring(self.element())

The returned string is of type 'str' for both python2 and python3.  Not sure if this is better or worse, up to you.

File: pandevice/base.py [hunk 7]
```diff
@@ -2775,10 +2781,10 @@ class XapiWrapper(pan.xapi.PanXapi):
         def __init__(self, *args, **kwargs):
             self.pan_device = kwargs.pop('pan_device', None)
             pan.xapi.PanXapi.__init__(self, *args, **kwargs)
-
+            pred = lambda x: inspect.ismethod(x) or inspect.isfunction(x)
```
Comment (btorresgil) [inline pandevice/base.py]: How come functions were added?  I think the only pan-python parts that are wrapped are methods, not functions.

File: .travis.yml [hunk 8]
```diff
@@ -5,6 +5,7 @@ language: python
 python:
   - "2.7"
   - "pypy"
+  - "3.6"
```
Comment (btorresgil) [inline .travis.yml]: Can you change the order to 2.7, 3.6, then pypy?

File: pandevice/__init__.py (hunk 3 above)
Comment (btorresgil) [inline pandevice/__init__.py]: Added a suggestion on another way to handle `element_str()`, since the `ElementTree.tostring()` method is capable of producing unicode natively in python 3.  See my comment here:  https://github.com/PaloAltoNetworks/pandevice/pull/59#discussion_r120527512

File: pandevice/base.py (hunk 7 above)
Comment (AdityaSripal) [inline pandevice/base.py]: Python3's inspect module has slightly different behavior with ismethod in python2. inspect.ismethod returns true for unbound and bound methods. In Python3 there is no notion of unbound methods so ismethod returns only bound methods. Perhaps the pan-python methods were unbound causing it to fail with Python3?
"Unbound methods" are functions according to Python3, but they are methods in python2.
Thus, I needed the pred function to get the tests to work with Python2 and Python3.
TL;DR:
inspect.ismethod works with Python2 but not Python3
inspect.isfunction works with Python3 but not Python2

File: tests/test_base.py [hunk 9]
```diff
@@ -12,14 +12,13 @@
 # ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 # OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 try:
-    import unittest.mock as mock
+    import unittest.mock
```
Comment (shinmog) [inline tests/test_base.py]: I think this needs to be `from unittest import mock`

File: tests/test_integration.py [hunk 10]
```diff
@@ -12,12 +12,10 @@
 # ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 # OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 try:
-    import unittest.mock as mock
+    import unittest.mock
```
Comment (shinmog) [inline tests/test_integration.py]: I think this needs to be `from unittest import mock`

File: tests/test_versioning.py [hunk 11]
```diff
@@ -1,12 +1,11 @@
 try:
-    import unittest.mock as mock
+    import unittest.mock
```
Comment (shinmog) [inline tests/test_versioning.py]: I think this needs to be `from unittest import mock`

Comment (AdityaSripal) [inline tests/test_versioning.py]: Fixed in all test_files in commit #f7e40b8

File: pandevice/base.py (hunk 6 above)
Comment (kevinsteves) [inline pandevice/base.py]: XML on PAN-OS is utf8.  suggest to just use encoding='utf8' for 2 and 3.  2 will be str, 3 will be bytes.  3 needs bytes for urlopen() so unicode would need to be encoded back to bytes by pan.xapi.

File: tests/test_integration.py (hunk 4 above)
Comment (shinmog) [inline tests/test_integration.py]: Yeah, please remove the `maxDiff`s since they don't do anything.

File: pandevice/__init__.py [hunk 12]
```diff
@@ -180,8 +211,9 @@ def string_or_list(value):
     """
     if value is None:
         return None
-    else:
-        return list(value) if "__iter__" in dir(value) else [value]
+    if isstring(value):
+        return [value,]
+    return list(value) if "__iter__" in dir(value) else [value]
```
Comment (shinmog) [inline pandevice/__init__.py]: Also, please make this `[value, ]` instead of `[value]`?

File: examples/userid.py [hunk 13]
```diff
@@ -106,7 +106,7 @@ def main():
         logging.debug("Logout user %s at IP %s" % (args.user, args.ip))
         device.userid.logout(args.user, args.ip)
     else:
-        raise StandardError("Unknown action: %s.  Must be 'login' or 'logout'." % args.action)
+        raise ValueError("Unknown action: %s.  Must be 'login' or 'logout'." % args.action) #Standard Error no longer supported
```
Comment (shinmog) [inline examples/userid.py]: Ok to remove this inline comment.

File: examples/ensure_security_rule.py [hunk 14]
```diff
@@ -27,12 +27,12 @@
...
 """
 
-import sys
+
 
 import pandevice
 import pandevice.firewall
 import pandevice.policies
-
+import sys
```
Comment (btorresgil) [inline examples/ensure_security_rule.py]: Agreed, built-in packages come first.  Recommend to undo this change.

File: pandevice/base.py (hunk 7 above)
Comment (btorresgil) [inline pandevice/base.py]: Got it, thanks!  Can you add a comment to the code indicating why it is handled this way?  Just a note about what is needed for python 2 and for python 3.  Thanks!

File: pandevice/base.py [hunk 15]
```diff
@@ -2993,7 +2992,7 @@ def op(self, cmd=None, vsys=None, xml=False, cmd_xml=True, extra_qs=None, retry_
         """
         element = self.xapi.op(cmd, vsys, cmd_xml, extra_qs, retry_on_peer=retry_on_peer)
         if xml:
-            return ET.tostring(element)
+            return ET.tostring(element, encoding='utf-8')
```
Comment (btorresgil) [inline pandevice/base.py]: I think `encoding='utf-8'` causes an exception on python 2.7.

Comment (shinmog) [inline pandevice/base.py]: It looks like it works..?  Where did you see this error?

```
In [1]: import xml.etree.ElementTree as ET

In [2]: e = ET.Element('something')

In [3]: e.text = 'serious'

In [4]: ET.tostring(e, encoding='utf-8')
Out[4]: '<something>serious</something>'
```

Comment (AdityaSripal) [inline pandevice/base.py]: All the tests pass on python 2.7.13

File: pandevice/base.py (hunk 6 above)
Comment (btorresgil) [inline pandevice/base.py]: Thanks Kevin, utf-8 is a great solution.

File: pandevice/base.py (hunk 15 above)
Comment (btorresgil) [inline pandevice/base.py]: Nevermind, it was `encoding='unicode'` that caused the exception in python 27.  This `utf-8` solution works great.  Please disregard my comment.

**Comment by btorresgil** [comment]
👍  Looks great, will merge shortly.  Thanks for all the hard work, Aditya!

**Comment by FliesLikeABrick** [comment]
Thank you to everyone involved in writing, proofing, and merging this code in!
